### PR TITLE
Modular function summaries

### PR DESCRIPTION
**Pull request by jerhard**
This is work in progress on implementing the modular function summaries. 

The (internal) project documenting the tasks can be found [here](https://github.com/orgs/goblint/projects/12)

### CODE REVIEW

File: src/analyses/usedGlobals.ml [hunk 1]
```diff
@@ -0,0 +1,132 @@
...
+  include Analyses.DefaultSpec
+
+  let name () = "used_globals"
+  module D = SetDomain.Make(CilType.Varinfo)
+  module C = Lattice.Unit
+
+  let context _ _ = ()
+
+  let collect_in_expression (exp: exp) (p : varinfo -> bool) : D.t =
+    (* TODO: Implement as visitor instead?*)
```
Comment (sim642) [inline src/analyses/usedGlobals.ml:17]: That would be shorter indeed with just `method! vvrbl`.

File: src/analyses/written.ml [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+        | _ -> failwith "map_back yielded a non-address value for address input."
+      in
+      let v' = ModularUtil.ValueDomainExtension.map_back v ~reachable in
+      D.add k' v' map
+    in
+    D.fold translate_and_insert au (ctx.local)
+
+  let combine_assign ctx (lval:lval option) fexp (f:fundec) (args:exp list) fc (au:D.t) (f_ask: Queries.ask) : D.t =
+    (* Record assignment of rval to lval*)
+    ctx.local
```
Comment (sim642) [inline src/analyses/written.ml:71]: Is this actually a TODO?

File: src/cdomains/valueDomain.ml [hunk 3]
```diff
@@ -197,6 +198,46 @@ struct
...
+      let init_field s fd =
+        let v, targets = top_value_typed_address_targets ~varAttr:fd.fattr fd.ftype in
+        (Structs.replace s fd v), targets
+      in
+      let init_field (s, acc) fd =
+        let v, targets = init_field s fd in
+        (v, targets @ acc)
+      in
+      let s, vs = List.fold_left init_field (Structs.top (), []) ci.cfields in
+      M.tracel "top_value_typed" "default_value for struct type %a is %a\n" d_type t Structs.pretty s;
```
Comment (sim642) [inline src/cdomains/valueDomain.ml:228]:
```suggestion
      if M.tracing then M.tracel "top_value_typed" "default_value for struct type %a is %a\n" d_type t Structs.pretty s;
```

File: src/cdomains/writtenDomain.ml [hunk 4]
```diff
@@ -0,0 +1,10 @@
+module VD = ValueDomain.Compound
+
+module AD = struct
+  include ValueDomain.AD
+  include Printable.Std
+end
```
Comment (sim642) [inline src/cdomains/writtenDomain.ml:6]: `Printable.Std` should always be included _first_ such that everything else can override those, instead of the dummy defaults overriding manual implementations:
```suggestion
module AD = struct
  include Printable.Std
  include ValueDomain.AD
end
```

File: src/domains/queries.ml [hunk 5]
```diff
@@ -9,6 +9,9 @@ module VDQ = ValueDomainQueries
 module ID = VDQ.ID
 
 module LS = VDQ.LS
+
+module AD = Lattice.LiftTop (ValueDomain.AD)
```
Comment (sim642) [inline src/domains/queries.ml:13]: Why does this need a `LiftTop`? `AD` already has a top.

File: src/maingoblint.ml [hunk 6]
```diff
@@ -133,6 +133,7 @@ let eprint_color m = eprintf "%s\n" (MessageUtil.colorize ~fd:Unix.stderr m)
 let check_arguments () =
   let fail m = (let m = "Option failure: " ^ m in eprint_color ("{red}"^m); failwith m) in(* unused now, but might be useful for future checks here *)
   let warn m = eprint_color ("{yellow}Option warning: "^m) in
+  if get_bool "modular" && get_bool "allfuns" then (set_bool "allfuns" false; warn "modular disables allfuns.\n");
```
Comment (sim642) [inline src/maingoblint.ml:136]: Maybe this should just be an error?

File: src/util/modularUtil0.ml [hunk 7]
```diff
@@ -0,0 +1,12 @@
...
+let type_to_varinfo t =
+  TypeVarinfoMap.to_varinfo t
+
+let varinfo_to_canonical x =
+  let t = x.vtype in
+  type_to_varinfo t
+
+(** If [x] is global and we are in ["modular"] mode, retrieves the varinfo_to_canonical for [x], else returns [x] itself *)
+let varinfo_or_canonical x =
+  if get_bool "modular" && x.vglob then varinfo_to_canonical x else x
```
Comment (sim642) [inline src/util/modularUtil0.ml:12]: This function probably gets called all the time around every `varinfo`, right? In that case it might later be worth caching the option value into a `bool ref` variable.

File: src/util/options.schema.json [hunk 8]
```diff
@@ -103,6 +103,12 @@
       "type": "boolean",
       "default": false
     },
+    "modular": {
+      "title": "modular",
+      "description": "Completlly modular analysis of all functions.",
```
Comment (sim642) [inline src/util/options.schema.json:108]:
```suggestion
      "description": "Completely modular analysis of all functions.",
```

File: src/util/typeVarinfoMap.ml [hunk 9]
```diff
@@ -0,0 +1,43 @@
...
+
+  let describe_varinfo (v: Cil.varinfo) (t: Cil.typ) =
+    TypeMap.describe_varinfo v t
+
+  let bindings () =
+    failwith "TypeVarinfoMap.bindings: not implemented"
+end
+
+(* TODO: Do not expose module above through second module path *)
+include TypeVarinfoMap
```
Comment (sim642) [inline src/util/typeVarinfoMap.ml:43]: There's no need to define an inner module at all then if it's just included, just make all those definitions top-level in the file.

File: src/util/typeVarinfoMap.ml [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+
+    let describe_varinfo v ts = ""
+  end
+
+  module TypeMap = RichVarinfo.BiVarinfoMap.Make(Typ)
+  include TypeMap
+  let to_varinfo (t: t) =
+    (* Invariant: Only sanitized types are inserted into TypeMap *)
+    let sanitize_type (t: Cil.typ) =
+      let t = unrollType t in
```
Comment (sim642) [inline src/util/typeVarinfoMap.ml:21]: Not sure what sanitized means here, but maybe it should be `unrollTypeDeep` then as well?

File: src/domains/queries.ml (hunk 5 above)
Comment (jerhard) [reply src/domains/queries.ml:13]: If it is not lifted, Goblint crashes with `Goblint_lib.Lattice.Unsupported("partial map top")`:

```
Fatal error: exception Goblint_lib.Lattice.Unsupported("partial map top")
Raised at Goblint_lib__Lattice.unsupported in file "src/domains/lattice.ml", line 49, characters 20-41
Called from Goblint_lib__MCP.MCP2.query' in file "src/analyses/mCP.ml", line 312, characters 37-52
Called from Goblint_lib__Analyses.ask_of_ctx.(fun) in file "src/framework/analyses.ml" (inlined), line 352, characters 81-90
Called from Goblint_lib__Written.Spec.assign in file "src/analyses/written.ml", line 26, characters 10-39
Called from Goblint_lib__MCP.MCP2.assign.f in file "src/analyses/mCP.ml", line 379, characters 17-34
Called from Goblint_lib__MCP.MCP2.map_deadcode.one_el in file "src/analyses/mCP.ml", line 100, characters 49-78
Called from Stdlib__List.fold_left in file "list.ml", line 121, characters 24-34
Called from Goblint_lib__MCP.MCP2.map_deadcode in file "src/analyses/mCP.ml", line 101, characters 13-35
Called from Goblint_lib__MCP.MCP2.assign in file "src/analyses/mCP.ml", line 381, characters 15-52
Called from Goblint_lib__Constraints.WidenContextLifterSide.lift_fun in file "src/framework/constraints.ml", line 373, characters 25-41
```

Comment (sim642) [reply src/domains/queries.ml:13]: Apparently there's this TODO:
https://github.com/goblint/analyzer/blob/17348bec5283822b54b6dd4503a8bb7b617a15ed/src/cdomains/addressDomain.ml#L146-L147

File: src/analyses/written.ml [hunk 11]
```diff
@@ -0,0 +1,154 @@
...
+  let startstate v = D.bot ()
+  let threadenter ctx ~multiple lval f args = [D.bot ()]
+  let threadspawn ctx ~multiple lval f args fctx = ctx.local
+  let exitstate  v = D.top ()
+
+  let query ctx (type a) (q: a Q.t): a Q.result =
+    match q with
+    | Written ->
+      let written : D.t = ctx.local in
+      written
```
Comment (michael-schwarz) [inline src/analyses/written.ml:143]:
```suggestion
      (ctx.local:D.t)
```

Does this also do the trick?

File: src/cdomain/value/cdomains/lval.ml [hunk 12]
```diff
@@ -1,3 +1,672 @@
 (** Domains for {!GoblintCil.lval}. *)
```
Comment (michael-schwarz) [inline src/cdomain/value/cdomains/lval.ml:1]: Where is all this commented out code coming from? Is this an artifact of a merge?

File: src/domains/queries.ml [hunk 13]
```diff
@@ -379,8 +424,22 @@ struct
...
+        let r = BaseDomain.CPA.compare c c2 in
+        if r <> 0 then
+          r
+        else begin
+          let r = AD.compare a1 a2 in
+          if r <> 0 then
+            r
+          else
+            AD.compare b1 b2
+        end
```
Comment (michael-schwarz) [inline src/domains/queries.ml:442]: `BatTuple.Tuple3.compare` ?

https://ocaml-batteries-team.github.io/batteries-included/hdoc2/BatTuple.Tuple3.html

File: src/domain/mapDomain.ml [hunk 14]
```diff
@@ -257,7 +257,7 @@ struct
 
   let relift = lift_f' M.relift
 end
-
+(*
```
Comment (michael-schwarz) [inline src/domain/mapDomain.ml:260]: Is this commented out because it no longer compiles?

File: src/domain/lattice.ml [hunk 15]
```diff
@@ -651,3 +663,43 @@ struct
...
+  let binop f x y = match x, y with
+    | Some x, Some y -> Some (f x y)
+    | None, Some x
+    | Some x, None -> Some x
+    | None, None -> None
+
+  let leq x y = match x, y with
+    | Some x, Some y -> Base.leq x y
+    | Some x, None -> false
+    | None, _ -> true
```
Comment (michael-schwarz) [inline src/domain/lattice.ml:681]: This seems to imply `None` is bot but then it really isn't? Is there any fix to get a proper lattice and still have it behave the way that is needed?

File: src/cdomain/value/cdomains/valueDomain.ml [hunk 16]
```diff
@@ -210,6 +211,65 @@ struct
     | TNamed ({ttype=t; _}, _) -> top_value ~varAttr t
     | _ -> Top
 
+  (* top values, except type-based target for addresses *)
```
Comment (michael-schwarz) [inline src/cdomain/value/cdomains/valueDomain.ml:214]: Can this be de-duplicated? E.g. by having a function and supplying it with different `top` functions for addresses?

File: src/cdomain/value/cdomains/unionDomain.ml [hunk 17]
```diff
@@ -6,12 +6,24 @@ module type Arg =
...
+  val get: fieldinfo -> t -> value
+  val get_field_and_value: t -> fieldinfo option * value
+  (** Return the field used in the last write of the union, and the value of that field *)
+
+  val replace: t -> fieldinfo -> value -> t
+  val map: (fieldinfo option -> value -> value) -> t -> t
+  val fold: (fieldinfo option -> value -> 'a -> 'a) -> t -> 'a -> 'a
+  val smart_leq: leq_elem: (value -> value -> bool) -> t -> t -> bool
+  val smart_join: join_elem:(value -> value -> value) -> t -> t -> t
+  val smart_widen: widen_elem:(value -> value -> value) -> t -> t -> t
```
Comment (michael-schwarz) [inline src/cdomain/value/cdomains/unionDomain.ml:25]: These `smart` operators seems very different from their smart counterparts for arrays. Maybe we should rename both to a better name?

File: src/analyses/timedAnalysis.ml [hunk 18]
```diff
@@ -0,0 +1,55 @@
...
+
+  let threadspawn ctx ~multiple lval f args fctx =
+    time "threadspawn" (threadspawn ~multiple ctx lval f args) fctx
+
+  let exitstate v =
+    time "exitstate" exitstate v
+end
+
+(* let _ =
+  MCP.register_analysis (module Spec : MCPSpec) *)
```
Comment (michael-schwarz) [inline src/analyses/timedAnalysis.ml:55]: Why not register it?  If it makes reviewing easier, it could be a separate PR and already merged into master?

File: src/analyses/readAnalysis.ml [hunk 19]
```diff
@@ -0,0 +1,161 @@
...
+module AD = Queries.AD
+module VD = ValueDomain.Compound
+
+module Spec =
+struct
+  include Analyses.DefaultSpec
+
+  let name () = "read"
+  module D = AD
+  module C = Lattice.Unit
```
Comment (michael-schwarz) [inline src/analyses/readAnalysis.ml:16]:
```suggestion
  module C = Printable.Unit
```

File: src/analyses/pthreadSignals.ml [hunk 20]
```diff
@@ -20,6 +20,10 @@ struct
   let possible_vinfos (a: Queries.ask) cv_arg =
     Queries.AD.to_var_may (a.f (Queries.MayPointTo cv_arg))
 
+  let is_modular ctx =
+    let ask = Analyses.ask_of_ctx ctx in
+    ask.f Queries.IsModular
+
```
Comment (michael-schwarz) [inline src/analyses/pthreadSignals.ml:26]: Maybe promote this to a helper all analyses can use?

File: src/analyses/modularWrapper.ml [hunk 21]
```diff
@@ -0,0 +1,296 @@
...
+
+  let modular_combine_env ctx lval fd exp f_ask =
+    map_ctx (fun ctx -> S.modular_combine_env ctx lval fd exp f_ask) ctx
+  let modular_combine_assign ctx lval fd exp f_ask =
+    map_ctx (fun ctx -> S.modular_combine_assign ctx lval fd exp f_ask) ctx
+
+end
+
+(* let _ =
+   MCP.register_analysis (module Spec : MCPSpec) *)
```
Comment (michael-schwarz) [inline src/analyses/modularWrapper.ml:296]: Does this no longer compile?

File: src/analyses/mallocWrapperAnalysis_ml
Comment (michael-schwarz) [inline src/analyses/mallocWrapperAnalysis_ml:1]: Committed by mistake? The file name has an underscore.

**Review by michael-schwarz — COMMENTED** [review]
I let some fairly mundane comments for now. I am happy to review it more in detail at some point when it's time for that, just ping me :smile: 
I have forgot many things again though, so i would appreciate a quick walk-through before such a more thorough review.

File: src/domain/mapDomain.ml (hunk 14 above)
Comment (jerhard) [reply src/domain/mapDomain.ml:260]: It no longer compiles because the signature that the Lattice.HConsed changed such that it allows projecting away parts of the state that are not needed for the modular mode during the analysis. 
The MapDomain.HConsed is also not used anywhere.

**Comment by michael-schwarz** [comment]
This is not something that is planned for the near-to mid future.
